Sources/SourceKitLSP/Swift/SyntacticTestIndex.swift

```diff
@@ -0,0 +1,225 @@
+  /// is older or the same as this date.
+  let sourceFileModificationDate: Date
+}
+
+/// Syntactically scans the file at the given URL for tests declared within it.
+///
+/// Does not write the results to the index.
+///
+/// The order of the returned tests is not defined. The results should be sorted before being returned to the editor.
+fileprivate func testItems(in url: URL) async -> [TestItem] {
+  guard url.pathExtension == "swift" else {
+    return []
+  }
+  let syntaxTreeManager = SyntaxTreeManager()
+  let snapshot = orLog("Getting document snapshot for swift-testing scanning") {
+    try DocumentSnapshot(withContentsFromDisk: url, language: .swift)
+  }
+  guard let snapshot else {
+    return []
+  }
+  async let swiftTestingTests = SyntacticSwiftTestingTestScanner.findTestSymbols(
+    in: snapshot,
+    syntaxTreeManager: syntaxTreeManager
+  )
+  async let xcTests = SyntacticSwiftXCTestScanner.findTestSymbols(in: snapshot, syntaxTreeManager: syntaxTreeManager)
+  return await swiftTestingTests + xcTests
+}
+
+actor SyntacticTestIndex {
+  /// The tests discovered by the index.
+  private var indexedTests: [DocumentURI: IndexedTests] = [:]
+
+  /// The queue on which the index is being updated and queried.
+  ///
+  /// Tracking dependencies between tasks within this queue allows us to start indexing tasks in parallel with low
+  /// priority and elevate their priority once a read task comes in, which has higher priority and depends on the
+  /// indexing tasks to finish.
+  private let indexingQueue = AsyncQueue<TaskMetadata>()
+
+  init() {}
+
+  private func removeFilesFromIndex(_ removedFiles: Set<DocumentURI>) {
+    // Cancel any tasks for the removed files to ensure any pending indexing tasks don't re-add index data for the
+    // removed files.
+    self.indexingQueue.cancelTasks(where: { taskMetadata in
+      guard case .index(let uri) = taskMetadata else {
+        return false
+      }
+      return removedFiles.contains(uri)
+    })
+    for removedFile in removedFiles {
+      self.indexedTests[removedFile] = nil
+    }
+  }
+
+  /// Called when the list of files that may contain tests is updated.
+  ///
+  /// All files that are not in the new list of test files will be removed from the index.
+  func listOfTestFilesDidChange(_ testFiles: [DocumentURI]) {
+    let removedFiles = Set(self.indexedTests.keys).subtracting(testFiles)
+    removeFilesFromIndex(removedFiles)
+
+    rescanFiles(testFiles)
+  }
+
+  func filesDidChange(_ events: [FileEvent]) {
+    for fileEvent in events {
+      switch fileEvent.type {
+      case .created:
+        // We don't know if this is a potential test file. It would need to be added to the index via
+        // `listOfTestFilesDidChange`
+        break
+      case .changed:
+        rescanFiles([fileEvent.uri])
+      case .deleted:
+        removeFilesFromIndex([fileEvent.uri])
```

Comment:
Oh, good catch. I didn’t think about that when changing to batch indexing. 

Modified it to explicitly keep track of removed files instead of relying on task cancellation.